stat/lda.go

```diff
@@ -0,0 +1,208 @@
+		}
+
+	}
+	//Create a new array with labels and go through the array of y values and if
+	//it doesnt exist then add it to the new array
+	sort.Ints(labels)
+	fmt.Printf("Sorted list of labels: %v \n", labels)
+
+	if labels[0] != 0 {
+		panic("Label does not start from zero")
+	}
+	for i := 0; i < len(labels); i++ {
+		if labels[i] < 0 {
+			panic("Negative class label")
+		}
+		if i > 0 && labels[i]-labels[i-1] > 1 {
+			panic("Missing class")
+		}
+	}
+	//Tol is a tolerence to decide if a covariance matrix is singular (det is zero)
+	//Tol will reject variables whose variance is less than tol
+	var tol float64 = 1E-4
+	//k is the number of classes
+	ld.k = len(labels)
+	fmt.Printf("this is k and ld.n: %v, %v \n", ld.k, ld.n)
+	if ld.k < 2 {
+		panic("Only one class.")
+	}
+	if tol < 0.0 {
+		panic("Invalid tol")
+	}
+	if ld.n <= ld.k {
+		panic("Sample size is too small")
+	}
+
+	//Number of instances in each class
+	ni := make([]int, ld.k)
+
+	//Common mean vector
+	var colmean []float64
+	for i := 0; i < ld.p; i++ {
+		var col []float64 = mat.Col(nil, i, x)
+		var sum float64 = 0
+		for _, value := range col {
+			sum += value
+		}
+		colmean = append(colmean, sum/float64(ld.n))
+	}
+	fmt.Printf("this is the array of means %v \n", colmean)
+
+	//C is a matrix of zeros with dimensions: ld.p x ld.p
+	C := mat.NewSymDense(ld.p, make([]float64, ld.p*ld.p, ld.p*ld.p))
+	fmt.Printf("this is the zero matrix: %v \n", C)
+
+	//Class mean vectors
+	//mu is a matrix with dimensions: k x ld.p
+	ld.mu = mat.NewDense(ld.k, ld.p, make([]float64, ld.k*ld.p, ld.k*ld.p))
+	for i := 0; i < ld.n; i++ {
+		ni[y[i]] = ni[y[i]] + 1
+		for j := 0; j < ld.p; j++ {
+			ld.mu.Set(y[i], j, ((ld.mu.At(y[i], j)) + (x.At(i, j))))
+		}
+	}
+	for i := 0; i < ld.k; i++ {
+		for j := 0; j < ld.p; j++ {
+			ld.mu.Set(i, j, ((ld.mu.At(i, j)) / (float64)(ni[i])))
+		}
+	}
+
+	//priori is the priori probability of each class
+	priori := make([]float64, ld.k)
+	for i := 0; i < ld.k; i++ {
+		priori[i] = (float64)(ni[i] / ld.n)
+	}
+
+	//ct is the constant term of discriminant function of each class
+	ld.ct = make([]float64, ld.k)
+	for i := 0; i < ld.k; i++ {
+		ld.ct[i] = math.Log(priori[i])
+	}
+
+	for i := 0; i < ld.n; i++ {
+		for j := 0; j < ld.p; j++ {
+			for l := 0; l <= j; l++ {
+				C.SetSym(j, l, (C.At(j, l) + ((x.At(i, j) - colmean[j]) * (x.At(i, l) - colmean[l]))))
+			}
+		}
+	}
+
+	tol = tol * tol
+
+	for j := 0; j < ld.p; j++ {
+		for l := 0; l <= j; l++ {
+			C.SetSym(j, l, ((C.At(j, l)) / (float64)(ld.n-ld.k)))
+			C.SetSym(l, j, C.At(j, l))
+		}
+		if C.At(j, j) < tol {
+			panic("Covarience matrix (variable %d) is close to singular")
+		}
+	}
+
+	fmt.Printf("this is the code varience %v \n", C)
+
+	//Factorize returns whether the decomposition succeeded
+	//If the decomposition failed, methods that require a successful factorization will panic
+	ld.eigen.Factorize(C, true)
+	fmt.Printf("this is the eigen value %v \n", ld.eigen)
+	return true
+}
```

Comment:
Rather than return `bool`, this function should probably return `error`.  It should be fairly easy to change: you'd change `return true` to `return nil`, and all the `panic`s would get changed to `return fmt.Errorf("Oh no - error description here!")`.

Code that calls it would look something like this:
```Go
if err := ld.LinearDiscriminent(x, y); err != nil {
  // something went wrong
  log.Printf("Error in LinearDiscriment: %v", err)
}
```